config/manifests/vllm/gpu-deployment.yaml

```diff
@@ -46,26 +46,93 @@ spec:
             - containerPort: 8000
               name: http
               protocol: TCP
+          lifecycle:
+            preStop:
+              # vLLM stops accepting connections when it receives SIGTERM, so we need to sleep
+              # to give upstream gateways a chance to take us out of rotation. The time we wait
+              # is dependent on the time it takes for all upstreams to completely remove us from
+              # rotation. Older or simpler load balancers might take upwards of 30s, but we expect
+              # our deployment to run behind a modern gateway like Envoy which is designed to 
+              # probe for readiness aggressively.
+              sleep:
+                # Upstream gateway probers for health should be set on a low period, such as 5s,
+                # and the shorter we can tighten that bound the faster that we release
+                # accelerators during controlled shutdowns. However, we should expect variance,
+                # as load balancers may have internal delays, and we don't want to drop requests
+                # normally, so we're often aiming to set this value to a p99 propagation latency
+                # of readiness -> load balancer taking backend out of rotation, not the average.
+                # 
+                # This value is generally stable and must often be experimentally determined on
+                # for a given load balancer and health check period. We set the value here to
+                # the highest value we observe on a supported load balancer, and we recommend
+                # tuning this value down and verifying no requests are dropped.
+                #
+                # If this value is updated, be sure to update terminationGracePeriodSeconds.
+                #
+                seconds: 30
```

Comment:
I am thinking this is not needed. 
1. vllm naturally use uvicorn to run the api server, which support denying new requests but wait until received requests done. there is a graceful timeout `timeout_graceful_shutdown` in it. But seems vllm does not  configure it, not sure is it equals to no timeout or immediately shutdown.
2. gateway can move the vllm instance being deleted by checking the deletiontimestamp. This is what istio does now

Reply:
vLLM in my testing by default continues to process existing requests after SIGTERM while denying new requests, which is one prereq for graceful shutdown.

The other prereq: Load balancers in front of terminating pods still send requests until the termination propagates to them and they can roll out that config.  In clouds or on-premise, that can be tens of seconds to minutes.  Even with an HA configuration of a very fast propagation (pod terminated -> endpoint updated -> gateway controller reacts -> distributed via xDS -> all replicas react) it could still be 5-10s.  This interval needs to be set to the longest time period that any supported gateway in front of the inference pool can react to propagation of endpoint data, which was 30s for GKE Gateway right now.

Reply:
I set to 30s because I expected most direct envoy implementations to respond faster, but they will still not be able to reliably propagate <5s (in large endpoint slices it might take the endpoint controller several seconds or more to react to pod timestamp being set).

Reply:
>The other prereq: Load balancers in front of terminating pods still send requests until the termination propagates to them and they can roll out that config.

Yes, this totally correct. An optional way should make gateway to handle fallback